app/components/windows/WidgetEditor.vue.ts

```diff
@@ -29,79 +33,81 @@ import { ToggleInput, NumberInput } from 'components/shared/inputs/inputs';
     GenericForm,
     VFormGroup,
     TestWidgets,
-    Display
+    Display,
+    CustomFieldsEditor,
+    CodeEditor
   }
 })
 export default class WidgetWindow extends Vue {
-  @Inject() private sourcesService: ISourcesServiceApi;
+  @Inject() private widgetsService: IWidgetsServiceApi;
   @Inject() private windowsService: WindowsService;
-  @Inject() private widgetsService: WidgetsService;
   @Inject() private projectorService: ProjectorService;
 
-  @Prop() slots: any[];
-  @Prop() settings: any[];
-  @Prop() value: boolean;
-  @Prop() requestState: 'fail' | 'success' | 'pending';
-  @Prop() loaded: boolean;
+  /**
+   * Declaration of additional sections in the right panel
+   * @see example of usage in TipJar.vue.ts
+   */
+  @Prop() slots?: IWidgetNavItem[];
+
+  /**
+   * Navigation items for the right panel
+   */
+  @Prop() navItems: IWidgetNavItem[];
 
   $refs: { content: HTMLElement, sidebar: HTMLElement, code: HTMLElement };
 
+  sourceId = this.windowsService.getChildWindowOptions().queryParams.sourceId;
+  widget = this.widgetsService.getWidgetSource(this.sourceId);
+  apiSettings = this.widget.getSettingsService().getApiSettings();
+  properties: TObsFormData = [];
   codeTabs = [
     { value: 'HTML', name: $t('HTML') },
     { value: 'CSS', name: $t('CSS') },
-    { value: 'JS', name: $t('JS') },
-    { value: 'customFields', name: $t('Custom Fields') }
+    { value: 'JS', name: $t('JS') }
   ];
-  canRender = false; // prevents window flickering
-  sourceId = this.windowsService.getChildWindowOptions().queryParams.sourceId;
-  source = this.sourcesService.getSource(this.sourceId);
-  widgetType = this.source.getPropertiesManagerSettings().widgetType;
-  widgetUrl = this.service.getPreviewUrl();
-  previewSource: ISourceApi = null;
   currentTopTab = 'editor';
   currentCodeTab = 'HTML';
   currentSetting = 'source';
-  properties: TObsFormData = [];
-  tabs: IWidgetTab[] = [];
-  tabsList: { name: string, value: string}[] = [];
-
-  sourceUpdatedSubscr: Subscription;
+  readonly settingsState = this.widget.getSettingsService().state;
 
-  get service() {
-    return this.widgetsService.getWidgetSettingsService(this.widgetType);
+  get loaded() {
+    return !!this.settingsState.data;
   }
 
   get loadingFailed() {
-    return this.requestState === 'fail' && !this.loaded;
+    return !this.loaded && this.settingsState.loadingState == 'fail';
+  }
+
+  get wData(): IWidgetData {
+    if (!this.settingsState.data) return null;
+    return cloneDeep(this.settingsState.data) as IWidgetData;
```

Comment:
if wData is just a clone here will it still update via the API when we enable custom code?

Reply:
Yes, since now all changes for widget settings are synced via websockets

Reply:
Just checked, that's fine here. The socket event calls a mutation so it will be reactive here